[PDI-4048] Text file input cannot properly handle a fixed-length-of-bytes file

HiromuHota opened the pull request:
Hi,

This patch adds a 'length' option so that a user can specify in which length ('Characters' or 'Bytes') a fixed-length file is formatted.

Currently, a file containing multi-byte characters can be properly parsed only when the length is character-based.
This issue occurs when a file contains multi-byte characters and each field has a fixed-length of bytes.
As stated by @Ex-S in the [Forum](http://forums.pentaho.com/archive/index.php/t-54882.html), this is not an unusual case since columns of a database table may be defined by bytes instead of characters and so are the exported fixed-length files.
Related issues are [PDI-6559](http://jira.pentaho.com/browse/PDI-6559) and [PDI-15469](http://jira.pentaho.com/browse/PDI-15469).

Regards,
Hiromu Hota

wingman-pentaho commented:
Hello, This is an automated comment from our build system. Since this pull request originated from a user  not in the Pentaho organization, we will not proceed  with automatic validation of the pull request.  However, rest assured that it will be reviewed manually as soon  as we get the chance. Thanks for your contribution!

wingman-pentaho commented:
### Build Completed

:x: This pull request has errors. They will need to be addressed before it can be accepted. See below for more details. Some links are also available below for further assistance in addressing those issues.

#### Build Commands

```sh
ant -Dpentaho.resolve.repo=http://ivy-nexus.pentaho.org/content/groups/approved -Dtestreports.dir=bin/reports/unit-test -f engine/build.xml clean-all resolve jacoco && ant -Dpentaho.resolve.repo=http://ivy-nexus.pentaho.org/content/groups/approved -Dtestreports.dir=bin/reports/integration-test -f engine/build.xml jacoco-integration checkstyle publish-local
ant -Dpentaho.resolve.repo=http://ivy-nexus.pentaho.org/content/groups/approved -Dtestreports.dir=bin/reports/unit-test -f ui/build.xml clean-all resolve jacoco && ant -Dpentaho.resolve.repo=http://ivy-nexus.pentaho.org/content/groups/approved -Dtestreports.dir=bin/reports/integration-test -f ui/build.xml jacoco-integration checkstyle publish-local
```

#### Cleanup Commands
```sh
rm -r ~/.ivy2/local || echo no publish local to remove
```

#### Changed files

```console
  engine/src/org/pentaho/di/trans/steps/fileinput/text/TextFileInputMeta.java
  engine/src/org/pentaho/di/trans/steps/fileinput/text/TextFileInputUtils.java
  engine/src/org/pentaho/di/trans/steps/fileinput/text/messages/messages_en_US.properties
  engine/test-src/org/pentaho/di/trans/steps/fileinput/text/TextFileInputContentParsingTest.java
  engine/test-src/org/pentaho/di/trans/steps/fileinput/text/TextFileInputMetaNewInjectionTest.java
  engine/test-src/org/pentaho/di/trans/steps/fileinput/text/files/fixed-length-bytes.txt
  ui/src/org/pentaho/di/ui/trans/steps/fileinput/text/TextFileInputDialog.java
```

#### Unit test coverage change

These statistics help you identify how your changes have affected the coverage of the following files. If a file is not in this list, then its coverage was not affected by your changes. To get some help interpreting these metrics, please refer to [Jacoco's documentation](http://eclemma.org/jacoco/trunk/doc/counters.html).

###### org.pentaho.di.core.auth.core.impl.DefaultAuthenticationConsumerFactory
 * Branch Change:  + 7.14%
 * Complexity Change:  + 7.69%

###### org.pentaho.di.repository.kdr.KettleDatabaseRepositoryCreationHelper
 * Branch Change:  + 1.64%
 * Complexity Change:  + .69%
 * Instruction Change:  + 2.43%
 * Line Change:  + 3.70%

###### org.pentaho.di.trans.steps.checksum.CheckSumMeta
 * Branch Change:  + 2.63%
 * Complexity Change:  + 2.86%
 * Instruction Change:  + .54%
 * Line Change:  + .54%

###### org.pentaho.di.trans.steps.combinationlookup.CombinationLookupMeta
 * Instruction Change:  + .04%

###### org.pentaho.di.trans.steps.csvinput.CsvInputMeta
 * Instruction Change:  + .05%

###### org.pentaho.di.trans.steps.dbproc.DBProcMeta
 * Instruction Change:  + .11%

###### org.pentaho.di.trans.steps.fileinput.text.TextFileInputMeta
 * Instruction Change:  + .30%
 * Line Change:  + .23%

###### org.pentaho.di.trans.steps.fileinput.text.TextFileInputReader
 * Branch Change:  + 1.20%
 * Complexity Change:  + 1.12%
 * Instruction Change:  + 1.05%
 * Line Change:  + .53%

###### org.pentaho.di.trans.steps.fileinput.text.TextFileInputUtils
 * Branch Change: **-.11%**:small_red_triangle_down:
 * Complexity Change: **-.33%**:small_red_triangle_down:
 * Instruction Change:  + .41%
 * Line Change:  + .25%

###### org.pentaho.di.trans.steps.groupby.GroupByMeta
 * Branch Change:  + 4.17%
 * Complexity Change:  + 4.21%
 * Instruction Change:  + .28%
 * Line Change:  + .75%

###### org.pentaho.di.trans.steps.ifnull.IfNullMeta
 * Branch Change:  + 2.00%
 * Complexity Change:  + 1.89%
 * Instruction Change:  + .17%

###### org.pentaho.di.trans.steps.insertupdate.InsertUpdateMeta
 * Instruction Change:  + .05%

###### org.pentaho.di.trans.steps.ldapinput.LDAPInputMeta
 * Branch Change:  + 1.61%
 * Complexity Change:  + .97%
 * Instruction Change:  + .09%

###### org.pentaho.di.trans.steps.memgroupby.MemoryGroupByMeta
 * Branch Change: **-1.18%**:small_red_triangle_down:
 * Complexity Change: **-1.35%**:small_red_triangle_down:

###### org.pentaho.di.trans.steps.sortedmerge.SortedMergeMeta
 * Instruction Change: **-.18%**:small_red_triangle_down:

###### org.pentaho.di.trans.steps.tableoutput.TableOutputMeta
 * Branch Change:  + .91%
 * Complexity Change:  + .82%

###### org.pentaho.di.trans.Trans
 * Branch Change: **-.10%**:small_red_triangle_down:
 * Complexity Change: **-.14%**:small_red_triangle_down:
 * Instruction Change: **-.09%**:small_red_triangle_down:
 * Line Change: **-.11%**:small_red_triangle_down:

###### org.pentaho.di.ui.trans.steps.fileinput.text.TextFileInputDialog
 * Branch Change: **-.01%**:small_red_triangle_down:
 * Complexity Change: **-.01%**:small_red_triangle_down:
 * Instruction Change: **-.01%**:small_red_triangle_down:
 * Line Change: **-.01%**:small_red_triangle_down:

#### Integration test coverage change

These statistics help you identify how your changes have affected the coverage of the following files. If a file is not in this list, then its coverage was not affected by your changes. To get some help interpreting these metrics, please refer to [Jacoco's documentation](http://eclemma.org/jacoco/trunk/doc/counters.html).

###### org.pentaho.di.trans.step.BaseStep
 * Branch Change: **-.29%**:small_red_triangle_down:
 * Complexity Change: **-.18%**:small_red_triangle_down:
 * Instruction Change: **-.13%**:small_red_triangle_down:
 * Line Change: **-.08%**:small_red_triangle_down:

#### License header violations

> :heavy_exclamation_mark: Copyright year is not the current year on engine/test-src/org/pentaho/di/trans/steps/fileinput/text/TextFileInputMetaNewInjectionTest.java. Found 2015. Was expecting 2016

#### Checkstyle violations

The code submitted is not complying with one or more of our guidelines. To perform these checks locally, [please use this tool](https://github.com/pentaho/pentaho-coding-standards).

###### /home/buildguy/project/ui/src/org/pentaho/di/ui/trans/steps/fileinput/text/TextFileInputDialog.java

> :heavy_exclamation_mark: 'if' child have incorrect indentation level 8, expected level should be 6. at Line: 2,282

mbatchelor commented:
Hello Hiromu-san,

Thank-you so much for your contribution. There are just a few issues with your submission though, and I'd like to work with you to get them corrected:

1- You changed the items being serialized (by adding Length), but you didn't update TextFileInputMetaLoadSaveTest to test for your new variable.

2- I ran wingman on your PR, and there were two violations - one on checkstyle (indenting) and the other on the copyright date in the class header (see above).

Can you please make the fixes and update this Pull Request so I can process it?

Thanks.

Marc

matthewtckr commented on ui/src/org/pentaho/di/ui/trans/steps/fileinput/text/TextFileInputDialog.java:
```diff
@@ -1491,20 +1495,41 @@ public void focusGained( org.eclipse.swt.events.FocusEvent e ) {
+    wLength = new CCombo( wContentComp, SWT.BORDER | SWT.READ_ONLY );
+    wLength.setText( BaseMessages.getString( PKG, "TextFileInputDialog.Length.Label" ) );
+    props.setLook( wLength );
+    wLength.add( "Characters" );
```
@HiromuHota

We missed this in the first review, but there is also a need to have the "Characters" and "Bytes" strings defined by message properties, so that it can be localized into other languages.  As an example, the [CheckSumDialog class](https://github.com/pentaho/pentaho-kettle/blob/6.1.0.1-R/ui/src/org/pentaho/di/ui/trans/steps/checksum/CheckSumDialog.java#L155-L160) does this for selecting the CheckSum Type.

HiromuHota commented:
Hello Marc and Matt,

Thank you for reviewing my request and gave me feedbacks.
I fixed license (year) and style (indenting), and added message properties for later localization.
Also I updated TextFileInputMetaLoadSaveTest to test the new variable "length," but I noticed that other variables might also need testing load/save.

If you are going to update the load/save test for other variables, I'm happy to do that.
Please show me a reference implementation.

Regards,
Hiromu Hota

mbatchelor commented:
@HiromuHota - Go to https://github.com/pentaho/pentaho-kettle/blob/master/engine/test-src/org/pentaho/di/trans/steps/textfileinput/TextFileInputMetaLoadSaveTest.java and observe the "commonAttributes" list at the top of the setup() method. Your new field name needs to be added to array list. Provided your getter/setter method names are standard ( getLength(), setLength( String value ) ) then that's all you need to do.

If your getter/setter method names are non-standard ( like hasHeader for a boolean instead of isHeader( ) or like setFileNameForTest( String value ) ) then you will need to add an entry into the getter/setter map as shown. Special object types are also handled by the load/save tester infrastructure with the typeValidators.

The LoadSaveTester framework tests the following for you automatically:
* Load/Save round trip through different repository and file types ( save as XML, load into repo, read from repo, etc).
* Clone

HiromuHota commented:
Thank you very much for your kind instruction, but unfortunately I got a little bit confused.

There are two implementations of the TextFileInput: one in org.pentaho.di.trans.steps.textfileinput, and another in org.pentaho.di.trans.steps.fileinput.text.
According to [this commit](https://github.com/pentaho/pentaho-kettle/commit/463c6d4357f2d5e7b7ba696ca71e3e21bffc14e2), the former one (textfileinput) has been deprecated and replaced with the latter one (fileinput.text). Please correct me if I'm wrong.
In the above comment, you suggest me to look at the LoadSaveTest for the old one (textfileinput/TextFileInputMetaLoadSaveTest.java).
Does this mean I should add the new variable "length" to the old implementation as well?

The new TextFileInput has the corresponding [LoadSaveTest](https://github.com/pentaho/pentaho-kettle/blob/master/engine/test-src/org/pentaho/di/trans/steps/fileinput/text/TextFileInputMetaLoadSaveTest.java), but the "commonAttributes" list has only three elements (once it had all the variables until [this commit](https://github.com/pentaho/pentaho-kettle/commit/ef4526c8e2a5a5486dfd114ce09569a32c516c44#diff-d4a203d4bc3913b7b190841038af2d27) deleted them)
I'm adding the new variable "length" to the new one. So I added it to the corresponding LoadSaveTest and found that many other variables were missing from the test.

wingman-pentaho commented:
### Build Failed

:x: Something went wrong while validating this pull request.


##### Stdout log
###### (last 100 lines)

```console
Config enriched: If no api token specified, use environment variable
Config enriched: Set the slackbot url
Config enriched: If github SourceControlType and null StatusUpdateType, use github
Config enriched: If github SourceControlType and StatusUpdaterType and null Organization, use the one from SourceRetriever
Config enriched: If github SourceControlType and StatusUpdaterType and null Repository, use the one from SourceRetriever
Config enriched: If github SourceControlType and StatusUpdaterType and null PullRequest, use the one from SourceRetriever
Config enriched: CheckstyleAnalyzer is enabled by default
Retrieving source with org.pentaho.build.buddy.bundles.source.github.SlackbotGithubSourceRetriever@2ba3c603
Git clone: [git, clone, --depth=10, --branch, master, https://[GITHUB TOKEN]@github.com/pentaho/pentaho-kettle.git, pentaho-kettle-base-2731-c67fda28-acec-4c30-a0ff-c605fcb39e0a]
Git fetch: [git, fetch, --depth=12, origin, pull/2731/head:pullRequest]
Git merge: [git, merge, --no-edit, --no-ff, -X, theirs, pullRequest]
```

#### Stderr log
###### (last 100 lines)

```console
Cloning into 'pentaho-kettle-base-2731-c67fda28-acec-4c30-a0ff-c605fcb39e0a'...
From https://github.com/pentaho/pentaho-kettle
 * [new ref]         refs/pull/2731/head -> pullRequest
error: Your local changes to the following files would be overwritten by merge:
	plugins/kettle-shapefilereader-plugin/src/org/pentaho/di/shapefilereader/ShapeFileReaderMeta.java
Please, commit your changes or stash them before you can merge.
Aborting
java.lang.Exception: java.io.IOException: java.io.IOException: Exit code: 128 for command: [git, merge, --no-edit, --no-ff, -X, theirs, pullRequest]
	at org.pentaho.build.buddy.bundles.orchestrator.OrchestratorImpl.orchestrate(OrchestratorImpl.java:379)
	at Proxycd885883_1d11_4b9a_93e8_df891da9b031.orchestrate(Unknown Source)
	at org.pentaho.build.buddy.bundles.rest.OrchestratorRestService$1.write(OrchestratorRestService.java:62)
	at org.apache.cxf.jaxrs.provider.BinaryDataProvider.writeTo(BinaryDataProvider.java:172)
	at org.apache.cxf.jaxrs.utils.JAXRSUtils.writeMessageBody(JAXRSUtils.java:1381)
	at org.apache.cxf.jaxrs.interceptor.JAXRSOutInterceptor.serializeMessage(JAXRSOutInterceptor.java:244)
	at org.apache.cxf.jaxrs.interceptor.JAXRSOutInterceptor.processResponse(JAXRSOutInterceptor.java:120)
	at org.apache.cxf.jaxrs.interceptor.JAXRSOutInterceptor.handleMessage(JAXRSOutInterceptor.java:83)
	at org.apache.cxf.phase.PhaseInterceptorChain.doIntercept(PhaseInterceptorChain.java:308)
	at org.apache.cxf.interceptor.OutgoingChainInterceptor.handleMessage(OutgoingChainInterceptor.java:83)
	at org.apache.cxf.phase.PhaseInterceptorChain.doIntercept(PhaseInterceptorChain.java:308)
	at org.apache.cxf.transport.ChainInitiationObserver.onMessage(ChainInitiationObserver.java:121)
	at org.apache.cxf.transport.http.AbstractHTTPDestination.invoke(AbstractHTTPDestination.java:253)
	at org.apache.cxf.transport.servlet.ServletController.invokeDestination(ServletController.java:234)
	at org.apache.cxf.transport.servlet.ServletController.invoke(ServletController.java:208)
	at org.apache.cxf.transport.servlet.ServletController.invoke(ServletController.java:160)
	at org.apache.cxf.transport.servlet.CXFNonSpringServlet.invoke(CXFNonSpringServlet.java:180)
	at org.apache.cxf.transport.servlet.AbstractHTTPServlet.handleRequest(AbstractHTTPServlet.java:298)
	at org.apache.cxf.transport.servlet.AbstractHTTPServlet.doPost(AbstractHTTPServlet.java:217)
	at javax.servlet.http.HttpServlet.service(HttpServlet.java:707)
	at org.apache.cxf.transport.servlet.AbstractHTTPServlet.service(AbstractHTTPServlet.java:273)
	at org.eclipse.jetty.servlet.ServletHolder.handle(ServletHolder.java:812)
	at org.eclipse.jetty.servlet.ServletHandler.doHandle(ServletHandler.java:587)
	at org.ops4j.pax.web.service.jetty.internal.HttpServiceServletHandler.doHandle(HttpServiceServletHandler.java:70)
	at org.eclipse.jetty.server.handler.ScopedHandler.handle(ScopedHandler.java:143)
	at org.eclipse.jetty.security.SecurityHandler.handle(SecurityHandler.java:577)
	at org.eclipse.jetty.server.session.SessionHandler.doHandle(SessionHandler.java:223)
	at org.eclipse.jetty.server.handler.ContextHandler.doHandle(ContextHandler.java:1127)
	at org.ops4j.pax.web.service.jetty.internal.HttpServiceContext.doHandle(HttpServiceContext.java:271)
	at org.eclipse.jetty.servlet.ServletHandler.doScope(ServletHandler.java:515)
	at org.eclipse.jetty.server.session.SessionHandler.doScope(SessionHandler.java:185)
	at org.eclipse.jetty.server.handler.ContextHandler.doScope(ContextHandler.java:1061)
	at org.eclipse.jetty.server.handler.ScopedHandler.handle(ScopedHandler.java:141)
	at org.ops4j.pax.web.service.jetty.internal.JettyServerHandlerCollection.handle(JettyServerHandlerCollection.java:80)
	at org.eclipse.jetty.server.handler.HandlerWrapper.handle(HandlerWrapper.java:97)
	at org.eclipse.jetty.server.Server.handle(Server.java:499)
	at org.eclipse.jetty.server.HttpChannel.handle(HttpChannel.java:311)
	at org.eclipse.jetty.server.HttpConnection.onFillable(HttpConnection.java:257)
	at org.eclipse.jetty.io.AbstractConnection$2.run(AbstractConnection.java:544)
	at org.eclipse.jetty.util.thread.QueuedThreadPool.runJob(QueuedThreadPool.java:635)
	at org.eclipse.jetty.util.thread.QueuedThreadPool$3.run(QueuedThreadPool.java:555)
	at java.lang.Thread.run(Thread.java:745)
Caused by: java.io.IOException: java.io.IOException: Exit code: 128 for command: [git, merge, --no-edit, --no-ff, -X, theirs, pullRequest]
	at org.pentaho.build.buddy.bundles.source.github.GithubSourceRetriever.retrieveSource(GithubSourceRetriever.java:132)
	at Proxy2a04fb01_0bb6_4a50_a902_eb3885f206e8.retrieveSource(Unknown Source)
	at org.pentaho.build.buddy.bundles.orchestrator.OrchestratorImpl.orchestrate(OrchestratorImpl.java:303)
	... 41 more
Caused by: java.io.IOException: Exit code: 128 for command: [git, merge, --no-edit, --no-ff, -X, theirs, pullRequest]
	at org.pentaho.build.buddy.util.shell.ShellUtil.executeAndCheck(ShellUtil.java:59)
	at org.pentaho.build.buddy.bundles.source.github.GithubSourceRetriever.retrieveSource(GithubSourceRetriever.java:117)
	... 43 more

```

lucboudreau commented:
@mbatchelor Marc, mind if we keep this one opened a bit? I'd like to understand why Wingman fails to do the merge of this PR's code. The file listed above was created after this PR's parent commit, and this seems to cause issues. I'll look into it.

wingman-pentaho commented:
### Build Failed

:x: Something went wrong while validating this pull request.


##### Stdout log
###### (last 100 lines)

```console
Config enriched: If no api token specified, use environment variable
Config enriched: Set the slackbot url
Config enriched: If github SourceControlType and null StatusUpdateType, use github
Config enriched: If github SourceControlType and StatusUpdaterType and null Organization, use the one from SourceRetriever
Config enriched: If github SourceControlType and StatusUpdaterType and null Repository, use the one from SourceRetriever
Config enriched: If github SourceControlType and StatusUpdaterType and null PullRequest, use the one from SourceRetriever
Config enriched: CheckstyleAnalyzer is enabled by default
Retrieving source with org.pentaho.build.buddy.bundles.source.github.SlackbotGithubSourceRetriever@2ba3c603
Git clone: [git, clone, --depth=10, --branch, master, https://[GITHUB TOKEN]@github.com/pentaho/pentaho-kettle.git, pentaho-kettle-base-2731-a310d0c5-78e4-4d0d-b11f-67a92414864c]
Git fetch: [git, fetch, --depth=12, origin, pull/2731/head:pullRequest]
Git merge: [git, merge, --no-edit, --no-ff, -X, theirs, pullRequest]
```

#### Stderr log
###### (last 100 lines)

```console
Cloning into 'pentaho-kettle-base-2731-a310d0c5-78e4-4d0d-b11f-67a92414864c'...
From https://github.com/pentaho/pentaho-kettle
 * [new ref]         refs/pull/2731/head -> pullRequest
error: Your local changes to the following files would be overwritten by merge:
	plugins/kettle-shapefilereader-plugin/src/org/pentaho/di/shapefilereader/ShapeFileReader.java
	plugins/kettle-shapefilereader-plugin/src/org/pentaho/di/shapefilereader/ShapeFileReaderData.java
	plugins/kettle-shapefilereader-plugin/src/org/pentaho/di/shapefilereader/ShapeFileReaderMeta.java
Please, commit your changes or stash them before you can merge.
Aborting
java.lang.Exception: java.io.IOException: java.io.IOException: Exit code: 128 for command: [git, merge, --no-edit, --no-ff, -X, theirs, pullRequest]
	at org.pentaho.build.buddy.bundles.orchestrator.OrchestratorImpl.orchestrate(OrchestratorImpl.java:379)
	at Proxycd885883_1d11_4b9a_93e8_df891da9b031.orchestrate(Unknown Source)
	at org.pentaho.build.buddy.bundles.rest.OrchestratorRestService$1.write(OrchestratorRestService.java:62)
	at org.apache.cxf.jaxrs.provider.BinaryDataProvider.writeTo(BinaryDataProvider.java:172)
	at org.apache.cxf.jaxrs.utils.JAXRSUtils.writeMessageBody(JAXRSUtils.java:1381)
	at org.apache.cxf.jaxrs.interceptor.JAXRSOutInterceptor.serializeMessage(JAXRSOutInterceptor.java:244)
	at org.apache.cxf.jaxrs.interceptor.JAXRSOutInterceptor.processResponse(JAXRSOutInterceptor.java:120)
	at org.apache.cxf.jaxrs.interceptor.JAXRSOutInterceptor.handleMessage(JAXRSOutInterceptor.java:83)
	at org.apache.cxf.phase.PhaseInterceptorChain.doIntercept(PhaseInterceptorChain.java:308)
	at org.apache.cxf.interceptor.OutgoingChainInterceptor.handleMessage(OutgoingChainInterceptor.java:83)
	at org.apache.cxf.phase.PhaseInterceptorChain.doIntercept(PhaseInterceptorChain.java:308)
	at org.apache.cxf.transport.ChainInitiationObserver.onMessage(ChainInitiationObserver.java:121)
	at org.apache.cxf.transport.http.AbstractHTTPDestination.invoke(AbstractHTTPDestination.java:253)
	at org.apache.cxf.transport.servlet.ServletController.invokeDestination(ServletController.java:234)
	at org.apache.cxf.transport.servlet.ServletController.invoke(ServletController.java:208)
	at org.apache.cxf.transport.servlet.ServletController.invoke(ServletController.java:160)
	at org.apache.cxf.transport.servlet.CXFNonSpringServlet.invoke(CXFNonSpringServlet.java:180)
	at org.apache.cxf.transport.servlet.AbstractHTTPServlet.handleRequest(AbstractHTTPServlet.java:298)
	at org.apache.cxf.transport.servlet.AbstractHTTPServlet.doPost(AbstractHTTPServlet.java:217)
	at javax.servlet.http.HttpServlet.service(HttpServlet.java:707)
	at org.apache.cxf.transport.servlet.AbstractHTTPServlet.service(AbstractHTTPServlet.java:273)
	at org.eclipse.jetty.servlet.ServletHolder.handle(ServletHolder.java:812)
	at org.eclipse.jetty.servlet.ServletHandler.doHandle(ServletHandler.java:587)
	at org.ops4j.pax.web.service.jetty.internal.HttpServiceServletHandler.doHandle(HttpServiceServletHandler.java:70)
	at org.eclipse.jetty.server.handler.ScopedHandler.handle(ScopedHandler.java:143)
	at org.eclipse.jetty.security.SecurityHandler.handle(SecurityHandler.java:577)
	at org.eclipse.jetty.server.session.SessionHandler.doHandle(SessionHandler.java:223)
	at org.eclipse.jetty.server.handler.ContextHandler.doHandle(ContextHandler.java:1127)
	at org.ops4j.pax.web.service.jetty.internal.HttpServiceContext.doHandle(HttpServiceContext.java:271)
	at org.eclipse.jetty.servlet.ServletHandler.doScope(ServletHandler.java:515)
	at org.eclipse.jetty.server.session.SessionHandler.doScope(SessionHandler.java:185)
	at org.eclipse.jetty.server.handler.ContextHandler.doScope(ContextHandler.java:1061)
	at org.eclipse.jetty.server.handler.ScopedHandler.handle(ScopedHandler.java:141)
	at org.ops4j.pax.web.service.jetty.internal.JettyServerHandlerCollection.handle(JettyServerHandlerCollection.java:80)
	at org.eclipse.jetty.server.handler.HandlerWrapper.handle(HandlerWrapper.java:97)
	at org.eclipse.jetty.server.Server.handle(Server.java:499)
	at org.eclipse.jetty.server.HttpChannel.handle(HttpChannel.java:311)
	at org.eclipse.jetty.server.HttpConnection.onFillable(HttpConnection.java:257)
	at org.eclipse.jetty.io.AbstractConnection$2.run(AbstractConnection.java:544)
	at org.eclipse.jetty.util.thread.QueuedThreadPool.runJob(QueuedThreadPool.java:635)
	at org.eclipse.jetty.util.thread.QueuedThreadPool$3.run(QueuedThreadPool.java:555)
	at java.lang.Thread.run(Thread.java:745)
Caused by: java.io.IOException: java.io.IOException: Exit code: 128 for command: [git, merge, --no-edit, --no-ff, -X, theirs, pullRequest]
	at org.pentaho.build.buddy.bundles.source.github.GithubSourceRetriever.retrieveSource(GithubSourceRetriever.java:132)
	at Proxy2a04fb01_0bb6_4a50_a902_eb3885f206e8.retrieveSource(Unknown Source)
	at org.pentaho.build.buddy.bundles.orchestrator.OrchestratorImpl.orchestrate(OrchestratorImpl.java:303)
	... 41 more
Caused by: java.io.IOException: Exit code: 128 for command: [git, merge, --no-edit, --no-ff, -X, theirs, pullRequest]
	at org.pentaho.build.buddy.util.shell.ShellUtil.executeAndCheck(ShellUtil.java:59)
	at org.pentaho.build.buddy.bundles.source.github.GithubSourceRetriever.retrieveSource(GithubSourceRetriever.java:117)
	... 43 more

```

wingman-pentaho commented:
### Build Failed

:x: Something went wrong while validating this pull request.


##### Stdout log
###### (last 100 lines)

```console
Config enriched: If no api token specified, use environment variable
Config enriched: Set the slackbot url
Config enriched: If github SourceControlType and null StatusUpdateType, use github
Config enriched: If github SourceControlType and StatusUpdaterType and null Organization, use the one from SourceRetriever
Config enriched: If github SourceControlType and StatusUpdaterType and null Repository, use the one from SourceRetriever
Config enriched: If github SourceControlType and StatusUpdaterType and null PullRequest, use the one from SourceRetriever
Config enriched: CheckstyleAnalyzer is enabled by default
Retrieving source with org.pentaho.build.buddy.bundles.source.github.SlackbotGithubSourceRetriever@3c135fb0
Git clone: [git, clone, --depth=10, --branch, master, https://[GITHUB TOKEN]@github.com/pentaho/pentaho-kettle.git, pentaho-kettle-base-2731-ec34d97f-a124-4a30-bba1-a8d9b40be30e]
Git fetch: [git, fetch, --depth=12, origin, pull/2731/head:pullRequest]
Git merge: [git, merge, --no-edit, --no-ff, -X, theirs, pullRequest]
```

#### Stderr log
###### (last 100 lines)

```console
Cloning into 'pentaho-kettle-base-2731-ec34d97f-a124-4a30-bba1-a8d9b40be30e'...
From https://github.com/pentaho/pentaho-kettle
 * [new ref]         refs/pull/2731/head -> pullRequest
error: Your local changes to the following files would be overwritten by merge:
	plugins/kettle-shapefilereader-plugin/src/org/pentaho/di/shapefilereader/ShapeFileReader.java
	plugins/kettle-shapefilereader-plugin/src/org/pentaho/di/shapefilereader/ShapeFileReaderData.java
	plugins/kettle-shapefilereader-plugin/src/org/pentaho/di/shapefilereader/ShapeFileReaderMeta.java
Please, commit your changes or stash them before you can merge.
Aborting
java.lang.Exception: java.io.IOException: java.io.IOException: Exit code: 128 for command: [git, merge, --no-edit, --no-ff, -X, theirs, pullRequest]
	at org.pentaho.build.buddy.bundles.orchestrator.OrchestratorImpl.orchestrate(OrchestratorImpl.java:385)
	at Proxyd4ff1c93_8940_460c_ab5d_0897405ec2b6.orchestrate(Unknown Source)
	at org.pentaho.build.buddy.bundles.rest.OrchestratorRestService$1.write(OrchestratorRestService.java:62)
	at org.apache.cxf.jaxrs.provider.BinaryDataProvider.writeTo(BinaryDataProvider.java:172)
	at org.apache.cxf.jaxrs.utils.JAXRSUtils.writeMessageBody(JAXRSUtils.java:1381)
	at org.apache.cxf.jaxrs.interceptor.JAXRSOutInterceptor.serializeMessage(JAXRSOutInterceptor.java:244)
	at org.apache.cxf.jaxrs.interceptor.JAXRSOutInterceptor.processResponse(JAXRSOutInterceptor.java:120)
	at org.apache.cxf.jaxrs.interceptor.JAXRSOutInterceptor.handleMessage(JAXRSOutInterceptor.java:83)
	at org.apache.cxf.phase.PhaseInterceptorChain.doIntercept(PhaseInterceptorChain.java:308)
	at org.apache.cxf.interceptor.OutgoingChainInterceptor.handleMessage(OutgoingChainInterceptor.java:83)
	at org.apache.cxf.phase.PhaseInterceptorChain.doIntercept(PhaseInterceptorChain.java:308)
	at org.apache.cxf.transport.ChainInitiationObserver.onMessage(ChainInitiationObserver.java:121)
	at org.apache.cxf.transport.http.AbstractHTTPDestination.invoke(AbstractHTTPDestination.java:253)
	at org.apache.cxf.transport.servlet.ServletController.invokeDestination(ServletController.java:234)
	at org.apache.cxf.transport.servlet.ServletController.invoke(ServletController.java:208)
	at org.apache.cxf.transport.servlet.ServletController.invoke(ServletController.java:160)
	at org.apache.cxf.transport.servlet.CXFNonSpringServlet.invoke(CXFNonSpringServlet.java:180)
	at org.apache.cxf.transport.servlet.AbstractHTTPServlet.handleRequest(AbstractHTTPServlet.java:298)
	at org.apache.cxf.transport.servlet.AbstractHTTPServlet.doPost(AbstractHTTPServlet.java:217)
	at javax.servlet.http.HttpServlet.service(HttpServlet.java:707)
	at org.apache.cxf.transport.servlet.AbstractHTTPServlet.service(AbstractHTTPServlet.java:273)
	at org.eclipse.jetty.servlet.ServletHolder.handle(ServletHolder.java:812)
	at org.eclipse.jetty.servlet.ServletHandler.doHandle(ServletHandler.java:587)
	at org.ops4j.pax.web.service.jetty.internal.HttpServiceServletHandler.doHandle(HttpServiceServletHandler.java:70)
	at org.eclipse.jetty.server.handler.ScopedHandler.handle(ScopedHandler.java:143)
	at org.eclipse.jetty.security.SecurityHandler.handle(SecurityHandler.java:577)
	at org.eclipse.jetty.server.session.SessionHandler.doHandle(SessionHandler.java:223)
	at org.eclipse.jetty.server.handler.ContextHandler.doHandle(ContextHandler.java:1127)
	at org.ops4j.pax.web.service.jetty.internal.HttpServiceContext.doHandle(HttpServiceContext.java:271)
	at org.eclipse.jetty.servlet.ServletHandler.doScope(ServletHandler.java:515)
	at org.eclipse.jetty.server.session.SessionHandler.doScope(SessionHandler.java:185)
	at org.eclipse.jetty.server.handler.ContextHandler.doScope(ContextHandler.java:1061)
	at org.eclipse.jetty.server.handler.ScopedHandler.handle(ScopedHandler.java:141)
	at org.ops4j.pax.web.service.jetty.internal.JettyServerHandlerCollection.handle(JettyServerHandlerCollection.java:80)
	at org.eclipse.jetty.server.handler.HandlerWrapper.handle(HandlerWrapper.java:97)
	at org.eclipse.jetty.server.Server.handle(Server.java:499)
	at org.eclipse.jetty.server.HttpChannel.handle(HttpChannel.java:311)
	at org.eclipse.jetty.server.HttpConnection.onFillable(HttpConnection.java:257)
	at org.eclipse.jetty.io.AbstractConnection$2.run(AbstractConnection.java:544)
	at org.eclipse.jetty.util.thread.QueuedThreadPool.runJob(QueuedThreadPool.java:635)
	at org.eclipse.jetty.util.thread.QueuedThreadPool$3.run(QueuedThreadPool.java:555)
	at java.lang.Thread.run(Thread.java:745)
Caused by: java.io.IOException: java.io.IOException: Exit code: 128 for command: [git, merge, --no-edit, --no-ff, -X, theirs, pullRequest]
	at org.pentaho.build.buddy.bundles.source.github.GithubSourceRetriever.retrieveSource(GithubSourceRetriever.java:166)
	at Proxyde54bd2d_f457_4d7c_afd7_07a4552c8240.retrieveSource(Unknown Source)
	at org.pentaho.build.buddy.bundles.orchestrator.OrchestratorImpl.orchestrate(OrchestratorImpl.java:303)
	... 41 more
Caused by: java.io.IOException: Exit code: 128 for command: [git, merge, --no-edit, --no-ff, -X, theirs, pullRequest]
	at org.pentaho.build.buddy.util.shell.ShellUtil.executeAndCheck(ShellUtil.java:59)
	at org.pentaho.build.buddy.bundles.source.github.GithubSourceRetriever.retrieveSource(GithubSourceRetriever.java:117)
	... 43 more

```

wingman-pentaho commented:
### Build Failed

:x: Something went wrong while validating this pull request.


##### Stdout log
###### (last 100 lines)

```console
Config enriched: If no api token specified, use environment variable
Config enriched: Set the slackbot url
Config enriched: If github SourceControlType and null StatusUpdateType, use github
Config enriched: If github SourceControlType and StatusUpdaterType and null Organization, use the one from SourceRetriever
Config enriched: If github SourceControlType and StatusUpdaterType and null Repository, use the one from SourceRetriever
Config enriched: If github SourceControlType and StatusUpdaterType and null PullRequest, use the one from SourceRetriever
Config enriched: CheckstyleAnalyzer is enabled by default
Retrieving source with org.pentaho.build.buddy.bundles.source.github.SlackbotGithubSourceRetriever@d2d659
Git clone: [git, clone, --depth=10, --branch, master, https://[GITHUB TOKEN]@github.com/pentaho/pentaho-kettle.git, pentaho-kettle-base-2731-be735b98-f98a-4307-a257-5a1e4df8cb1f]
Git fetch: [git, fetch, --depth=12, origin, pull/2731/head:pullRequest]
Git merge: [git, merge, --no-edit, --no-ff, -Xtheirs, -Xignore-space-change, pullRequest]
```

#### Stderr log
###### (last 100 lines)

```console
Cloning into 'pentaho-kettle-base-2731-be735b98-f98a-4307-a257-5a1e4df8cb1f'...
From https://github.com/pentaho/pentaho-kettle
 * [new ref]         refs/pull/2731/head -> pullRequest
error: Your local changes to the following files would be overwritten by merge:
	plugins/kettle-shapefilereader-plugin/src/org/pentaho/di/shapefilereader/ShapeFileReader.java
	plugins/kettle-shapefilereader-plugin/src/org/pentaho/di/shapefilereader/ShapeFileReaderData.java
	plugins/kettle-shapefilereader-plugin/src/org/pentaho/di/shapefilereader/ShapeFileReaderMeta.java
Please, commit your changes or stash them before you can merge.
Aborting
java.lang.Exception: java.io.IOException: java.io.IOException: Exit code: 128 for command: [git, merge, --no-edit, --no-ff, -Xtheirs, -Xignore-space-change, pullRequest]
	at org.pentaho.build.buddy.bundles.orchestrator.OrchestratorImpl.orchestrate(OrchestratorImpl.java:385)
	at Proxyfa88e04e_3f90_4998_8300_17f71e41bce7.orchestrate(Unknown Source)
	at org.pentaho.build.buddy.bundles.rest.OrchestratorRestService$1.write(OrchestratorRestService.java:62)
	at org.apache.cxf.jaxrs.provider.BinaryDataProvider.writeTo(BinaryDataProvider.java:172)
	at org.apache.cxf.jaxrs.utils.JAXRSUtils.writeMessageBody(JAXRSUtils.java:1381)
	at org.apache.cxf.jaxrs.interceptor.JAXRSOutInterceptor.serializeMessage(JAXRSOutInterceptor.java:244)
	at org.apache.cxf.jaxrs.interceptor.JAXRSOutInterceptor.processResponse(JAXRSOutInterceptor.java:120)
	at org.apache.cxf.jaxrs.interceptor.JAXRSOutInterceptor.handleMessage(JAXRSOutInterceptor.java:83)
	at org.apache.cxf.phase.PhaseInterceptorChain.doIntercept(PhaseInterceptorChain.java:308)
	at org.apache.cxf.interceptor.OutgoingChainInterceptor.handleMessage(OutgoingChainInterceptor.java:83)
	at org.apache.cxf.phase.PhaseInterceptorChain.doIntercept(PhaseInterceptorChain.java:308)
	at org.apache.cxf.transport.ChainInitiationObserver.onMessage(ChainInitiationObserver.java:121)
	at org.apache.cxf.transport.http.AbstractHTTPDestination.invoke(AbstractHTTPDestination.java:253)
	at org.apache.cxf.transport.servlet.ServletController.invokeDestination(ServletController.java:234)
	at org.apache.cxf.transport.servlet.ServletController.invoke(ServletController.java:208)
	at org.apache.cxf.transport.servlet.ServletController.invoke(ServletController.java:160)
	at org.apache.cxf.transport.servlet.CXFNonSpringServlet.invoke(CXFNonSpringServlet.java:180)
	at org.apache.cxf.transport.servlet.AbstractHTTPServlet.handleRequest(AbstractHTTPServlet.java:298)
	at org.apache.cxf.transport.servlet.AbstractHTTPServlet.doPost(AbstractHTTPServlet.java:217)
	at javax.servlet.http.HttpServlet.service(HttpServlet.java:707)
	at org.apache.cxf.transport.servlet.AbstractHTTPServlet.service(AbstractHTTPServlet.java:273)
	at org.eclipse.jetty.servlet.ServletHolder.handle(ServletHolder.java:812)
	at org.eclipse.jetty.servlet.ServletHandler.doHandle(ServletHandler.java:587)
	at org.ops4j.pax.web.service.jetty.internal.HttpServiceServletHandler.doHandle(HttpServiceServletHandler.java:70)
	at org.eclipse.jetty.server.handler.ScopedHandler.handle(ScopedHandler.java:143)
	at org.eclipse.jetty.security.SecurityHandler.handle(SecurityHandler.java:577)
	at org.eclipse.jetty.server.session.SessionHandler.doHandle(SessionHandler.java:223)
	at org.eclipse.jetty.server.handler.ContextHandler.doHandle(ContextHandler.java:1127)
	at org.ops4j.pax.web.service.jetty.internal.HttpServiceContext.doHandle(HttpServiceContext.java:271)
	at org.eclipse.jetty.servlet.ServletHandler.doScope(ServletHandler.java:515)
	at org.eclipse.jetty.server.session.SessionHandler.doScope(SessionHandler.java:185)
	at org.eclipse.jetty.server.handler.ContextHandler.doScope(ContextHandler.java:1061)
	at org.eclipse.jetty.server.handler.ScopedHandler.handle(ScopedHandler.java:141)
	at org.ops4j.pax.web.service.jetty.internal.JettyServerHandlerCollection.handle(JettyServerHandlerCollection.java:80)
	at org.eclipse.jetty.server.handler.HandlerWrapper.handle(HandlerWrapper.java:97)
	at org.eclipse.jetty.server.Server.handle(Server.java:499)
	at org.eclipse.jetty.server.HttpChannel.handle(HttpChannel.java:311)
	at org.eclipse.jetty.server.HttpConnection.onFillable(HttpConnection.java:257)
	at org.eclipse.jetty.io.AbstractConnection$2.run(AbstractConnection.java:544)
	at org.eclipse.jetty.util.thread.QueuedThreadPool.runJob(QueuedThreadPool.java:635)
	at org.eclipse.jetty.util.thread.QueuedThreadPool$3.run(QueuedThreadPool.java:555)
	at java.lang.Thread.run(Thread.java:745)
Caused by: java.io.IOException: java.io.IOException: Exit code: 128 for command: [git, merge, --no-edit, --no-ff, -Xtheirs, -Xignore-space-change, pullRequest]
	at org.pentaho.build.buddy.bundles.source.github.GithubSourceRetriever.retrieveSource(GithubSourceRetriever.java:166)
	at Proxy4718ef8b_cc9a_4cc0_9988_72471e760c74.retrieveSource(Unknown Source)
	at org.pentaho.build.buddy.bundles.orchestrator.OrchestratorImpl.orchestrate(OrchestratorImpl.java:303)
	... 41 more
Caused by: java.io.IOException: Exit code: 128 for command: [git, merge, --no-edit, --no-ff, -Xtheirs, -Xignore-space-change, pullRequest]
	at org.pentaho.build.buddy.util.shell.ShellUtil.executeAndCheck(ShellUtil.java:59)
	at org.pentaho.build.buddy.bundles.source.github.GithubSourceRetriever.retrieveSource(GithubSourceRetriever.java:117)
	... 43 more

```

lucboudreau commented:
@mbatchelor Hi Marc, I've discovered that the Kettle project [overrides the Git behavior](https://github.com/pentaho/pentaho-kettle/blob/master/.gitattributes) and forces the client to make line endings conversions.

This means that:

- We clone the repository locally.
- The git client changes all line endings.
- We merge the pull request over the original repo.
- The merge fails because the file was modified locally.

I've tried all sorts of changes to the git command line to try and override this, but I have not found a way yet. Sorry, but this PR won't be able to be validated until we figure this out.

HiromuHota commented:
Thank you for all your efforts on this.
If there is anything I can do, please let me know.

mbatchelor commented:
Closing in favor of reworked #3340. Thank-you for the contribution, sorry we couldn't just merge this one.

HiromuHota commented:
Thank you.
